### PR TITLE
New example: weather demo

### PR DESCRIPTION
**Pull request by Justyna-JustCode**
Hi,
Please let me know if something needs to be added.

I assumed that links like:
https://slint.dev/snapshots/master/demos/weather-demo/
or
https://slint.dev/snapshots/master/editor?load_url=https://raw.githubusercontent.com/slint-ui/slint/master/examples/weather-demo/ui/main.slint
would start working when the PR is merged into the master and the wasm is built with the GitHub actions.
Let me know if I'm wrong.
Anyway, the preview with slint-pad is not working because we're using changed callbacks in the demo (see [here](https://snapshots.slint.dev/master/editor/?load_url=https://raw.githubusercontent.com/justyna-justcode/slint/feature/weather-demo/examples/weather-demo/ui/main.slint)). Should I remove the link, then?

### CODE REVIEW

**Comment by tronical** [comment]
> Anyway, the preview with slint-pad is not working because we're using changed callbacks in the demo (see [here](https://snapshots.slint.dev/master/editor/?load_url=https://raw.githubusercontent.com/justyna-justcode/slint/feature/weather-demo/examples/weather-demo/ui/main.slint)). Should I remove the link, then?

While change callbacks are still experimental and not enabled by default, we'd like to avoid their use in examples and demos (that folks copy & paste from). Would it be possible to change the code to not use them?

**Comment by tronical** [comment]
> I assumed that links like: https://slint.dev/snapshots/master/demos/weather-demo/ or https://slint.dev/snapshots/master/editor?load_url=https://raw.githubusercontent.com/slint-ui/slint/master/examples/weather-demo/ui/main.slint would start working when the PR is merged into the master and the wasm is built with the GitHub actions. Let me know if I'm wrong.

Yes, they should. I can help fix that up afterwards if necessary, no worries :)

File: examples/weather-demo/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+log = "0.4.21"
+serde = { version = "1.0", features = ["derive"] }
+serde_json = "1.0.115"
+
+[target.'cfg(all(not(target_os = "windows"), not(target_os = "android")))'.dependencies]
+slint = { path = "../../api/rs/slint" }
+
+# Using different Slint configuration on Windows to disable the accessibility feature (enabled by default).
+# The feature caused performance issues on Windows (see https://github.com/slint-ui/slint/issues/5338).
+# Note: might be fixed with v1.7
```
Comment (tronical) [inline examples/weather-demo/Cargo.toml:21]: Since this is being built with "1.7", does this still apply?

File: examples/weather-demo/build.rs [hunk 2]
```diff
@@ -0,0 +1,7 @@
+use std::env;
+
+fn main() {
+    env::set_var("SLINT_ENABLE_EXPERIMENTAL_FEATURES", "true");
```
Comment (tronical) [inline examples/weather-demo/build.rs:4]: I'm hesitant about the use of change callbacks in the demos/examples. @ogoffart what's your opinion?

**Comment by tronical** [comment]
The demo looks great overall, btw! Thanks for contributing it.

Please see the message from GitHub about the mismatch between GitHub user and email address. You'll also see that our CI complains about copyright headers and SPDX license tags.

File: examples/weather-demo/build.rs (hunk 2 above)
Comment (Justyna-JustCode) [reply examples/weather-demo/build.rs:4]: Initially, we tried to avoid them. I believe there are at least one or two use cases where I could not find a workaround for not using them and still providing the functionality.
Maybe I'm missing something. Please let me know if you want me to examine this again and provide you with concrete examples.

File: examples/weather-demo/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+tokio = "1.37.0"
+
+[target.'cfg(target_arch = "wasm32")'.dependencies]
+console_log = "1.0"
+console_error_panic_hook = "0.1.7"
+wasm-bindgen = "0.2"
+wasm-bindgen-futures = "0.4"
+
+[build-dependencies]
+slint-build = "1.6"
```
Comment (tronical) [inline examples/weather-demo/Cargo.toml:44]: You'll need this to compile:

```suggestion
slint-build = { path = "../../api/rs/build" }
```

Comment (Justyna-JustCode) [reply examples/weather-demo/Cargo.toml:44]: Right, I missed this one! Thanks :slightly_smiling_face:

File: examples/weather-demo/Cargo.toml (hunk 1 above)
Comment (Justyna-JustCode) [reply examples/weather-demo/Cargo.toml:21]: I believe so since the ticket ([the one that mine was a duplicate of](https://github.com/slint-ui/slint/issues/5041)) is still open. But I can double-check this.

Comment (tronical) [reply examples/weather-demo/Cargo.toml:21]: Ok :). We can deal with this separately and leave this.

Comment (tronical) [reply examples/weather-demo/Cargo.toml:26]: Ok, I checked and this is fixed now in master AFAICS.

File: examples/weather-demo/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+[target.'cfg(target_os = "windows")'.dependencies]
+slint = { path = "../../api/rs/slint", default-features = false, features = [ "compat-1-2", "backend-winit", "renderer-femtovg" ] }
+
+[target.'cfg(all(not(target_arch = "wasm32"), not(target_os = "android")))'.dependencies]
+env_logger = "0.11.3"
+
+[target.'cfg(target_os = "android")'.dependencies]
+android_logger = "0.14.1"
+openssl = { version = "0.10", features = ["vendored"] }
+slint = { path = "../../api/rs/slint", features = [ "backend-android-activity-06" ] }
```
Comment (tronical) [inline examples/weather-demo/Cargo.toml:36]:
```suggestion
slint = { path = "../../api/rs/slint", features = [ "backend-android-activity-06" ] }

[target.'cfg(all(not(target_arch = "wasm32"), not(target_os = "android")))'.dependencies]
env_logger = "0.11.3"

[target.'cfg(target_os = "android")'.dependencies]
android_logger = "0.14.1"
openssl = { version = "0.10", features = ["vendored"] }
```

File: examples/weather-demo/ui/assets/font-awesome.ttf
Comment (tronical) [inline examples/weather-demo/ui/assets/font-awesome.ttf:1]: This file is rather large. Is this needed for some icons? Would it be possible to extract them as .svgs instead? We use icons from font-awesome in a few other place and opted into shipping smaller .svg files instead of the whole ttf.

File: examples/weather-demo/ui/assets/weathericons-font.ttf
Comment (tronical) [inline examples/weather-demo/ui/assets/weathericons-font.ttf:1]: Same applies here, it would be great to replace the 100kb with just the .svg files needed. Do you have an overview over what you're using from the file? (roughly :-)

File: examples/weather-demo/ui/assets/font-awesome.ttf
Comment (Justyna-JustCode) [reply examples/weather-demo/ui/assets/font-awesome.ttf:1]: Ok, I can change this. From what I remember, they are only a few usages.

File: examples/weather-demo/ui/assets/weathericons-font.ttf
Comment (Justyna-JustCode) [reply examples/weather-demo/ui/assets/weathericons-font.ttf:1]: Mostly the icons used for the weather ([see](https://github.com/slint-ui/slint/pull/5601/files#diff-71980582608fae59323128c4874f32c857b222610a786002fc606091456ed3a7R18)) + 4 icons for the day time ([see](https://github.com/slint-ui/slint/pull/5601/files#diff-7f687e7076c92db9c907e3944c3f2d6fbc11318f1af59e732f01aa2ec3a91542R121)) and maybe 1,2 more.
 I guess I can extract them as well.

File: examples/weather-demo/build.rs (hunk 2 above)
Comment (tronical) [reply examples/weather-demo/build.rs:7]: Alright, let's keep the use of the change handlers as they are :-)

File: examples/weather-demo/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,73 @@
+# Copyright © SixtyFPS GmbH <info@slint.dev>
+# SPDX-License-Identifier: MIT
+
+[package]
+name = "rusty-weather"
+version = "1.0.0"
```
Comment (tronical) [inline examples/weather-demo/Cargo.toml:6]: I think we usually match the version with the rest of Slint in the repo :)

```suggestion
version = "1.7.0"
```

File: examples/weather-demo/src/weather/openweathercontroller.rs [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+    CityData, CityWeatherData, DayWeatherData, ForecastWeatherData, GeoLocationData,
+    PrecipitationData, TemperatureData, WeatherCondition, WeatherController, WeatherData,
+};
+
+#[cfg(target_os = "android")]
+use std::env;
+
+const CITIES_STORED_FILE_NAME: &str = "cities_data.json";
+const ORGANIZATION_QUALIFIER: &str = "com"; // have to match android app name in cargo.toml
+const ORGANIZATION_NAME: &str = "felgo.demos"; // have to match android app name in cargo.toml
```
Comment (tronical) [inline examples/weather-demo/src/weather/openweathercontroller.rs:30]: Maybe this needs changing/adapting? Just a drive-by thought.

File: examples/weather-demo/README.md [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+ * using `async` features with a Slint-based application and combining it with the Tokio runtime.
+
+<br />
+<p>
+  <img src="docs/img/desktop-preview.png" height="415px" />
+  <img src="docs/img/android-preview.png" height="415px" hspace="5" />
+</p>
+
+## Weather data
+To enable real weather data from the [OpenWeather](https://openweathermap.org/) API, you must provide the `RUSTY_WEATHER_API_KEY` environment variable with your API key at build time. The [OpenCall API](https://openweathermap.org/price#onecall) subscription is required.
```
Comment (tronical) [inline examples/weather-demo/README.md:29]: This thought is beyond the scope of this PR: We have similar needs in the energy monitor demo, where we use weatherapi.com and we've got an API key in the CI set, so that the demos built have weather data support. It might make sense to unify these, i.e. either to switch the energy monitor over to use openweathermap.org or the other way around. The key from weatherapi.com is free.

File: examples/weather-demo/Cargo.toml (hunk 1 above)
Comment (Justyna-JustCode) [reply examples/weather-demo/Cargo.toml:26]: I checked and can confirm that the issue is gone. Let me update the file.

File: examples/weather-demo/README.md (hunk 7 above)
Comment (Justyna-JustCode) [reply examples/weather-demo/README.md:29]: That sounds like a good idea :slightly_smiling_face: 

Just so you know, the OpenWeather API is also mostly free. Only the OneCall API we used requires a paid subscription, and it is only paid for calls above 1000 per day.

The other APIs available for free keys can provide similar functionalities. We only switched to the OneCall version because we were facing issues. However, as we later discovered, those were issues in the crate we used rather than in the API itself.

File: examples/weather-demo/ui/assets/weathericons-font.ttf
Comment (Justyna-JustCode) [reply examples/weather-demo/ui/assets/weathericons-font.ttf:1]: I decided to keep the weather icons as is.
Two reasons:
1. The actual gain is small (~25K), while the effort to change this in code is not marginal.
2. Other weather providers might require a different set of icons.

Do you think it's okay?

Comment (tronical) [reply examples/weather-demo/ui/assets/weathericons-font.ttf:1]: Alright, sounds okay :)

File: examples/weather-demo/src/weather/openweathercontroller.rs (hunk 6 above)
Comment (Justyna-JustCode) [reply examples/weather-demo/src/weather/openweathercontroller.rs:30]: I guess it makes sense, but let me double-check this and get back to you :slightly_smiling_face:

**Comment by Justyna-JustCode** [comment]
Regarding the "reuse" specification, I'm still waiting for confirmation on our internal images licensing. Anything else should already be covered.

File: examples/weather-demo/Cargo.toml (hunk 4 above)
Comment (Justyna-JustCode) [reply examples/weather-demo/Cargo.toml]: Done

File: examples/weather-demo/ui/assets/font-awesome.ttf
Comment (Justyna-JustCode) [reply examples/weather-demo/ui/assets/font-awesome.ttf:1]: Done

File: examples/weather-demo/Cargo.toml (hunk 5 above)
Comment (Justyna-JustCode) [reply examples/weather-demo/Cargo.toml]: Done

File: examples/weather-demo/src/weather/openweathercontroller.rs (hunk 6 above)
Comment (Justyna-JustCode) [reply examples/weather-demo/src/weather/openweathercontroller.rs]: Sorry for the delay. I was offline last week.
I changed this. Please let me know if the new version is okay with you.

**Comment by Justyna-JustCode** [comment]
I think it's all done now; I synced with the latest master, and the CI is passing.
Please let me know if you need something else from our side :slightly_smiling_face:

**Review by tronical — APPROVED** [review]
Awesome, thank you!
